src/libsync/propagatecommonzsync.cpp

```diff
@@ -0,0 +1,362 @@
+ * parameters.
+ */
+int ZsyncGenerateRunnable::fcopy_hashes(FILE *fin, FILE *fout, size_t rsum_bytes, size_t hash_bytes)
+{
+    unsigned char buf[CHECKSUM_SIZE + 4];
+    size_t len;
+
+    while ((len = fread(buf, 1, sizeof(buf), fin)) > 0) {
+        /* write trailing rsum_bytes of the rsum (trailing because the second part of the rsum is more useful in practice for hashing), and leading checksum_bytes of the checksum */
+        if (fwrite(buf + 4 - rsum_bytes, 1, rsum_bytes, fout) < rsum_bytes)
+            break;
+        if (fwrite(buf + 4, 1, hash_bytes, fout) < hash_bytes)
+            break;
+    }
+    if (ferror(fin)) {
+        return stream_error("fread", fin);
+    }
+    if (ferror(fout)) {
+        return stream_error("fwrite", fout);
+    }
+
+    return 0;
+}
+
+/**
+ * Given one block of data, calculate the checksums for this block and write
+ * them (as raw bytes) to the given output stream
+ */
+int ZsyncGenerateRunnable::write_block_sums(unsigned char *buf, size_t got, FILE *f)
+{
+    struct rsum r;
+    unsigned char checksum[CHECKSUM_SIZE];
+
+    /* Pad for our checksum, if this is a short last block  */
+    if (got < _blocksize)
+        memset(buf + got, 0, _blocksize - got);
+
+    /* Do rsum and checksum, and convert to network endian */
+    r = rcksum_calc_rsum_block(buf, _blocksize);
+    rcksum_calc_checksum(&checksum[0], buf, _blocksize);
+    r.a = htons(r.a);
+    r.b = htons(r.b);
+
+    /* Write them raw to the stream */
+    if (fwrite(&r, sizeof r, 1, f) != 1)
+        return stream_error("fwrite", f);
+    if (fwrite(checksum, sizeof checksum, 1, f) != 1)
+        return stream_error("fwrite", f);
+
+    return 0;
+}
+
+/**
+ * Reads the data stream and writes to the zsync stream the blocksums for the
+ * given data. No compression handling.
+ */
+int ZsyncGenerateRunnable::read_stream_write_blocksums(FILE *fin, FILE *fout)
+{
+    unsigned char *buf = (unsigned char *)malloc(_blocksize);
+
+    if (!buf) {
+        fprintf(stderr, "out of memory\n");
+        exit(1);
+    }
+
+    while (!feof(fin)) {
+        int got = fread(buf, 1, _blocksize, fin);
+
+        if (got > 0) {
+            /* The SHA-1 sum, unlike our internal block-based sums, is on the whole file and nothing else - no padding */
+            SHA1Update(&_shactx, buf, got);
+
+            write_block_sums(buf, got, fout);
+            _len += got;
+        } else {
+            if (ferror(fin))
+                return stream_error("fread", fin);
+        }
+    }
+    free(buf);
+    return 0;
+}
+
+void ZsyncGenerateRunnable::run()
+{
+    // Create a temporary file to use with zsync_begin()
+    QTemporaryFile zsynctf, zsyncmeta;
+    zsyncmeta.open();
+    zsyncmeta.setAutoRemove(false);
+    zsynctf.open();
+
+    int metaHandle = zsyncmeta.handle();
+    int tfHandle = zsynctf.handle();
+
+    zsync_unique_ptr<FILE> meta(fdopen(dup(metaHandle), "w"), [](FILE *f) {
```

Comment:
Why are you not using QFile here, and everywhere you use raw file operation?
QFile has a constructor that opens a file descriptor as well.
I'd think this would lead to more readable code.

Reply:
These functions were copied from zsync source as-is since they are not part of the exported api. Needed when generating metadata for a file. My main task was to integrate zsync, not re-write it.

Reply:
Also, please see the archive of all the comments I linked, none of this is *new*.